modules/library/main/src/test/java/org/geotools/data/ows/URLCheckersTest.java

```diff
@@ -72,18 +74,99 @@ public void testMultipleCheckers() {
         MockURLChecker gsChecker =
                 new MockURLChecker(
                         "enabled", true, s -> s.matches("https?://www.geoserver.org/.*")) {};
+
+        MockURLChecker apiChecker =
+                new MockURLChecker(
+                        "enabled",
+                        true,
+                        s -> s.matches("^https?://localhost:8080/geoserver/ows\\?.*$")) {};
+
+        MockURLChecker owsChecker =
+                new MockURLChecker(
+                        "enabled",
+                        true,
+                        s ->
+                                s.matches(
+                                        "^https?://localhost:8080/geoserver/(\\w+/)?ows((\\?\\!\\\\.\\\\./).)*(\\\\?.*)?$")) {};
+
         URLCheckers.register(fileChecker);
         URLCheckers.register(gtChecker);
         URLCheckers.register(gsChecker);
+        URLCheckers.register(apiChecker);
+        URLCheckers.register(owsChecker);
 
         // invalid references
         assertThrows(URLCheckerException.class, () -> URLCheckers.confirm("http://google.com"));
         assertThrows(URLCheckerException.class, () -> URLCheckers.confirm("http://nyt.com"));
         assertThrows(URLCheckerException.class, () -> URLCheckers.confirm("file:///tmp"));
+        assertThrows(
+                URLCheckerException.class,
+                () -> URLCheckers.confirm("http://localhost/geoserver/styles/grass_mark.png"));
+        assertThrows(
+                URLCheckerException.class,
+                () -> URLCheckers.confirm("http://localhost:8080/geoserver/ows/.."));
 
         // valid references
         URLCheckers.confirm("https://www.geoserver.org/logo.png");
         URLCheckers.confirm("http://www.geotools.org/sld.xsd");
         URLCheckers.confirm("file:///data/dem.tif");
+        URLCheckers.confirm("http://localhost:8080/geoserver/ows?");
+        URLCheckers.confirm("http://localhost:8080/geoserver/styles/../ows?");
+        URLCheckers.confirm(
+                "http://localhost:8080/geoserver/ows?SERVICE=WMS&REQUEST=GetCapabilities&VERSION=1.3.0");
+        URLCheckers.confirm(
+                "https://localhost:8080/geoserver/ne/ows?SERVICE=WFS&REQUEST=GetCapabilities&VERSION=1.1.0");
+
+        // valid normalization
+        URLCheckers.confirm("file:///data/../data/dem.tif");
+        URLCheckers.confirm("file:/data/../data/dem.tif");
+        URLCheckers.confirm("file:///data/./dem.tif");
+        URLCheckers.confirm("file:/data/./dem.tif");
+
+        // invalid normalization
+        assertThrows(URLCheckerException.class, () -> URLCheckers.confirm("file:///tmp"));
```

Comment:
Maybe also add one to confirm path traversal has been neutralized, e.g. ``file:///data/../tmp``

Reply:
Added